sumpy/fmm.py

```diff
@@ -263,14 +323,44 @@ def __init__(self, code_container, queue, tree, dtype, fmm_level_to_order,
         self.extra_kwargs = source_extra_kwargs.copy()
         self.extra_kwargs.update(self.kernel_extra_kwargs)
 
-    # {{{ data vector utilities
+        if base_kernel.is_translation_invariant:
+            if translation_classes_data is None:
+                from warnings import warn
+                warn(
+                     "List 2 (multipole-to-local) translations will be "
+                     "unoptimized. Supply a translation_classes_data argument to "
+                     "the wrangler for optimized List 2.",
+                     SumpyTranslationClassesDataNotSuppliedWarning,
+
+                     stacklevel=2)
+                self.supports_optimized_m2l = False
+            else:
+                self.supports_optimized_m2l = True
+        else:
+            self.supports_optimized_m2l = False
+
+        self.translation_classes_data = translation_classes_data
 
+    # {{{ data vector utilities
     def _expansions_level_starts(self, order_to_size):
+        return self.build_csr_level_starts(order_to_size,
+                self.tree.level_start_box_nrs)
+
+    def build_csr_level_starts(self, order_to_size, level_starts):
```

Comment:
Make a separate function, to clarify that this isn't part of the class interface?

Reply:
Fixed in 3bca4dc0736363c3abe3babdec6f3de6515cc297